### ISSUE TITLE
CRA instances in Kubernetes fail to reconnect

### ISSUE THREAD

**Issue by ArcadeMode** [microsoft/CRA#27]
Hi! 

I've been using CRA as a base to build a stream processor and am running into a strange problem in kubernetes. When instances are starting up all works well, connections are set up and are functional. However it gets interesting when introducing failures.

By default there seems no support for detecting a dropped TCP connection in CRA so I built a simple ping-pong/keepalive based timeout mechanism to detect a broken connection on either side.

Steps to reproduce:
1. Create two vertices with a one way connection.
2. Spin up the kubernetes cluster with those vertices in it
3. Delete the pod of the sending vertex (using kubectl for example)
4. Kubernetes will recreate the deleted pod and it will restart.

Now this is where things get funky: both ends get reconnected, however the connection does not work, i.e. times out again and gets reconnected, and times out again etc..

My suspicion is that the changed IP of the killed vertex does not reflect in the living vertex. So it reconnect to the killed vertex's old IP while the killed vertex connects to the correct IP but gets ignored for some reason?

Please let me know if any clarification is needed or how I can circumvent this issue.

Thanks!

**Comment by ArcadeMode** [comment]
Interestingly it does seem to work correctly when instead of deleting a pod to scale the statefulset/deployment down to 0 instances 🤔

(and scale back up to 1 instance after that)

**Comment by badrishc** [comment]
Thanks for the clear description of the situation. I think I have a fix, will put out a PR for you to test out shortly :)

Edit: hmm, doesnt seem to be an issue where i thought.

**Comment by badrishc** [comment]
> By default there seems no support for detecting a dropped TCP connection in CRA 

When a TCP connection is dropped, the other side that is waiting for data from the stream (e.g., await stream.ReadBytes(..., cancellationToken)) will get an exception. You need the token. If you don't catch this exception (so CRA receives it), or rethrow it, then CRA will re-establish that connection. Is this not working for you?

**Comment by badrishc** [comment]
> the changed IP of the killed vertex does not reflect in the living vertex

We recheck Azure table for the latest version of the vertex connection info (IP/port) when the connection tries to get re-established so I don't see how this might happen.

I tried two instances talking to each other locally. Killed and restarted the first instance using a different port. That worked fine. Can you re-try with the latest in master? There is one bugfix I checked in (due to an error that crept in in the last PR).

**Comment by badrishc** [comment]
Also, you can check and see if the `cravertextable` on Azure has the latest correct IP address for the restarted vertex.

**Comment by ArcadeMode** [comment]
Hi @badrishc thanks for your replies!

> > By default there seems no support for detecting a dropped TCP connection in CRA
> 
> When a TCP connection is dropped, the other side that is waiting for data from the stream (e.g., await stream.ReadBytes(..., cancellationToken)) will get an exception. You need the token. If you don't catch this exception (so CRA receives it), or rethrow it, then CRA will re-establish that connection. Is this not working for you?

This works in some cases, typically the receiving end picks up a failure of the sender but the sender may not always observe a failure of the receiver.  [This blog](https://blog.stephencleary.com/2009/05/detection-of-half-open-dropped.html) was very helpfull in wrapping my head around the concept. I ended up adding simple keepalive messages as described in the blog. I do throw exceptions back to CRA under these circumstances:
- The program is unable to write to the stream (IOException)
- The keepalive message timeout is reached (IOException)
- CRA requests cancellation using the provided token (OperationCancelledException)
After this happens the connection does indeed get re-established, the strange part is that when sending messages over this connection they do not arrive at the receiver end (even though the receiver also "is connected").

> > the changed IP of the killed vertex does not reflect in the living vertex
> 
> We recheck Azure table for the latest version of the vertex connection info (IP/port) when the connection tries to get re-established so I don't see how this might happen.
> 
> I tried two instances talking to each other locally. Killed and restarted the first instance using a different port. That worked fine. Can you re-try with the latest in master? There is one bugfix I checked in (due to an error that crept in in the last PR).

> Also, you can check and see if the `cravertextable` on Azure has the latest correct IP address for the restarted vertex.

I will try to build a container with the latest version on master and see if that makes a difference, I actually am running a slightly older version so this may just be the issue. The table does indeed get updated with the new IP adres of the vertex that restarted so I suppose that is not the problem.

I will be back with observations after trying a newer CRA version for the workers.

**Comment by ArcadeMode** [comment]
So after creating a CRA worker container using the latest version the problem appeared to persist. I decided to attempt making a minimal reproducable sample to share here. However, I found that the problem did not occur in that sample. After a tedious bug-hunt I found that a buffering layer in my stream processor was the culprit!

Turns out CRA was working fine all along, my apologies for misplacing this issue here. 

PS. So far I greatly enjoy the features CRA provides, particularly combined with kubernetes it takes a lot of work off my hands!
